File: Python/Statement.py
'''File name: Statement.py
Created by Joshua Brock on 10/23/2018'''


from ArithmeticExpression import *
from ArithmeticOperator import ArithmeticOperator
from BooleanExpression import BooleanExpression
from Iter import Iter
from Block import Block


class Statement:
    '''An interface to be expanded upon by the following files:
IfStatement
PrintStatement
ForStatement
WhileStatement
AssignStatement'''


class IfStatement(Statement):
    def __init__(self, bin_expr, block1, block2):
        '''Precondition: bin_expr, block1, and block2 are of valid types
        Postcondition: default constructor - creates a new IfStatement'''
        assert (isinstance(bin_expr, BooleanExpression) and isinstance(block1, Block) and isinstance(block2, Block))

        if bin_expr is None:
            raise Exception('Null binary expression')
        if block1 is None or block2 is None:
            raise Exception('Null expression argument(s)')

        self.bin_expr = bin_expr
        self.block1 = block1
        self.block2 = block2

    def execute(self):
        '''Postcondition: if bin_expr is true, executes block1, otherwise executes block2'''
        if self.bin_expr.execute():
            self.block1.execute()
        else:
            self.block2.execute()


class PrintStatement(Statement):
    def __init__(self, arith_expr):
        '''Precondition: arith_expr is of valid type
        Postcondition: default constructor - creates a new PrintStatement'''
        assert arith_expr, ArithmeticExpression
        self.arith_expr = arith_expr

    def execute(self):
        '''Postcondition: prints out arith_expr in a string format'''
        if self.arith_expr is None:
            print("")
        else:
            print(self.arith_expr.evaluate())


class ForStatement(Statement):
    def __init__(self, id, iter, block):
        '''Precondition: id, iter, and block are of valid types
        Postcondition: default constructor - creates a new ForStatement'''
        assert (isinstance(id, Id) and isinstance(iter, Iter) and isinstance(block, Block))
        if block is None:
            raise Exception('Null expression argument')
        self.id = id
        self.iter = iter
        self.block = block

    def execute(self):
        '''Postcondition: assigns id to iter's start value, then executes block. id's value is in/decremented
            Should exit loop once id is equal to iter's end value, otherwise we risk an infinite loop'''
        AssignStatement(self.id, self.iter.arithbegin).execute()

        if self.id.evaluate() < self.iter.arithend.evaluate():
            changeidval = BinaryExpression(ArithmeticOperator.ADD_OP, self.id, Constant(1))
        else:
            changeidval = BinaryExpression(ArithmeticOperator.SUB_OP, self.id, Constant(1))

        while self.id.evaluate() != self.iter.arithend.evaluate():
            self.block.execute()
            AssignStatement(self.id, changeidval).execute()


class WhileStatement(Statement):
    def __init__(self, bool_expr, block):
        '''Precondition: bool_expr and block are of valid types
        Postcondition: default constructor - creates a new WhileStatement'''
        assert (isinstance(bool_expr, BooleanExpression) and isinstance(block, Block))
        if bool_expr is None:
            raise Exception('Null boolean argument')
        if block is None:
            raise Exception('Null expression argument')
        self.bool_expr = bool_expr
        self.block = block

    def execute(self):
        '''Postcondition: executes block while bool_expr remains true
        WARNING: block should contribute to bool_expr becoming false else infinite loop can occur'''
        while self.bool_expr.execute():
            self.block.execute()


class AssignStatement(Statement):
    def __init__(self, var, expr):
        '''Precondition: var and expr are of valid types
        Postcondition: default constructor - creates a new AssignStatement'''
        assert (isinstance(var, Id) and isinstance(expr, ArithmeticExpression))
        if var is None:
            raise Exception('Null ID argument')
        if expr is None:
            raise Exception('Null expression argument')
        self.var = var
        self.expr = expr

    def execute(self):
        '''Postcondition: stores expr's value in memory alongside var's character'''
        store(self.var.ch, self.expr.evaluate())

File: CSharp/Statement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TextBNFConverter
{
    public interface Statement
    {
        /*
		    An interface to be expanded upon by the following classes:
		    IfStatement
		    PrintStatement
		    ForStatement
		    WhileStatement
		    AssignStatement
	    */
        void execute();
    }

    public class IfStatement : Statement
    {
        //IfStatement takes two (2) Block objects and a BoolExpression. Only one (1) block will be ran at any time based on the result of the BoolExpression
        private Block block1, block2;
        private BoolExpression bool_expr;
        public IfStatement(Block block1, Block block2, BoolExpression bool_expr)
        {
            //Postcondition: Creates new IfStatement using two (2) Block objects and one (1) BoolExpression
            if (block1 == null || block2 == null)
                throw new Exception("Null Block inputted into new IfStatement");
            this.block1 = block1;
            this.block2 = block2;
            this.bool_expr = bool_expr ?? throw new Exception("Null BoolExpression inputted into new IfStatement");
        }
        public void execute()
        {
            //Postcondition: If bool_expr is true, execute block1, otherwise execute block2
            if (bool_expr.execute())
                block1.execute();
            else
                block2.execute();
        }
    }
    public class PrintStatement : Statement
    {
        //PrintStatement merely prints an ArithExpression's value to the console
        private ArithExpression arith;
        public PrintStatement(ArithExpression arith)
        {
            this.arith = arith ?? throw new Exception("Null ArithExpression inputted to new PrintStatement");
        }
        public void execute()
        {
            //Postcondition: Prints the ArithExpression's value to the console
            Console.WriteLine($"{arith.evaluate()}");
        }
    }
    public class ForStatement : Statement
    {
        //ForStatement takes a Block, Iter, and Id. Iter is the Id's start and ending values. Executes Block every time Id's value modifies
        private Block block;
        private Iter iter;
        private Id id;
        public ForStatement(Block block, Iter iter, Id id)
        {
            this.block = block ?? throw new Exception("Null Block inputted into new ForStatement");
            this.iter = iter ?? throw new Exception("Null Iter inputted into new ForStatement");
            this.id = id ?? throw new Exception("Null Id inputted into new ForStatement");
        }
        public void execute()
        {
            //Postcondition: Executes Block continuously as Id modifies from its start to its end value.
            new AssignStatement(id, iter.getBegin).execute();

            //Decides whether our Id needs to increment or decrement to reach the end value
            BinaryExpression changeidval;
            if (id.evaluate() < iter.getEnd.evaluate())
                changeidval = new BinaryExpression(id, new Constant(1), ArithOperator.ADD_OP);
            else
                changeidval = new BinaryExpression(id, new Constant(1), ArithOperator.SUB_OP);

            //Execution conditions
            while (id.evaluate() != iter.getEnd.evaluate())
            {
                block.execute();
                new AssignStatement(id, changeidval).execute();
            }
        }
    }
    public class WhileStatement : Statement
    {
        //WhileStatement takes a BoolExpression and a Block. It repeatedly executes Block until BoolExpression returns false
        private BoolExpression bool_expr;
        private Block block;
        public WhileStatement(BoolExpression bool_expr, Block block)
        {
            //Postcondition: Creates a new WhileStatement using BoolExpression and Block
            this.bool_expr = bool_expr ?? throw new Exception("Null BoolExpression inputted into new WhileStatement");
            this.block = block ?? throw new Exception("Null Block inputted into new WhileStatement");
        }
        public void execute()
        {
            //Postcondition: Executes Block continuously until BoolExpression returns false
            while (bool_expr.execute())
                block.execute();
        }
    }
    public class AssignStatement : Statement
    {
        //AssignStatement takes an Id and an ArithExpression. It sets the Id equal to the ArithExpression's value (i.e. [Id] = [ArithExpression]) 
        private Id id;
        private ArithExpression arith;
        public AssignStatement(Id id, ArithExpression arith)
        {
            this.id = id ?? throw new Exception("Null Id inputted into new AssignStatement");
            this.arith = arith ?? throw new Exception("Null ArithExpression inputted into new AssignStatement");
        }
        public void execute()
        {
            //Postcondition: Makes the Id's current value equal to the ArithExpression's value
            Memory.store(id.Character, arith.evaluate());
        }
    }
}
